File: LeetCode/remote/spiral_matrix.rb
# https://leetcode.com/submissions/detail/37492731/
#
# HAHAHAHAHAHA!!! 
#
# You are here!
# Your runtime beats 100.00% of ruby coders.
#
# 22 / 22 test cases passed.
#   Status: Accepted
#   Runtime: 72 ms
#       
#       Submitted: 0 minutes ago
#

def spiral_order(matrix)
    offsets = [[ 0,  1], # right first
               [ 1,  0], # down second
               [ 0, -1], # left third
               [-1,  0]  # up forth
              ].cycle
    
    result = []
    return result if matrix.nil? || matrix.size == 0
    start = [0, 0]
    offset = offsets.next
    (matrix.size * matrix[0].size).times {
        result << matrix[start.first][start.last]
        matrix[start.first][start.last] = nil
        next_start = [start.first + offset.first, start.last + offset.last]
        offset = offsets.next if !next_start.first.between?(0, matrix.size - 1) ||
                                 !next_start.last.between?(0, matrix[0].size - 1) ||
                                 matrix[next_start.first][next_start.last] == nil
        start = [start.first + offset.first, start.last + offset.last]
    }

    return result
end

puts spiral_order([
                   [ 1, 2, 3 ],
                   [ 4, 5, 6 ],
                   [ 7, 8, 9 ]
                  ]).inspect

puts spiral_order([
                   [ 1, 2, 3, 4 ],
                   [12,13,14, 5 ],
                   [11,16,15, 6 ],
                   [10, 9, 8, 7 ]
                  ]).inspect

puts spiral_order([
                   [ 1, 2, 3, 4 ]
                  ]).inspect

File: LeetCode/remote/spiral_matrix.cs
//  https://leetcode.com/problems/spiral-matrix/
//
//  Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order. 
//

//  https://leetcode.com/submissions/detail/52900246/
//
//
//  Submission Details
//  22 / 22 test cases passed.
//      Status: Accepted
//      Runtime: 568 ms
//          
//          Submitted: 0 minutes ago

public class Solution {
    public IList<int> SpiralOrder(int[,] matrix) {
        var start = Tuple.Create(0, -1);
        var offset = GetOffset(new [] {
            Tuple.Create(0, 1),
            Tuple.Create(1, 0),
            Tuple.Create(0, -1),
            Tuple.Create(-1, 0)
        }).GetEnumerator();
        offset.MoveNext();
        return matrix.GetLength(0) == 0 || matrix.GetLength(1) == 0 ?
               new List<int>() :
               Enumerable
               .Range(0, matrix.GetLength(0) * matrix.GetLength(1))
               .Select(x => {
                   var next = Tuple.Create(start.Item1 + offset.Current.Item1,
                                           start.Item2 + offset.Current.Item2);
 
                   if (0 > next.Item1 || next.Item1 >= matrix.GetLength(0) ||
                       0 > next.Item2 || next.Item2 >= matrix.GetLength(1) ||
                       matrix[next.Item1, next.Item2] == int.MaxValue)
                       {
                           offset.MoveNext();
                           next = Tuple.Create(start.Item1 + offset.Current.Item1,
                                               start.Item2 + offset.Current.Item2);
                       }
                       
                   var result = matrix[next.Item1, next.Item2];
                   start = next;
                   matrix[next.Item1, next.Item2] = int.MaxValue;
                   return result;
               })
               .ToList();
    }
    
    public IEnumerable<Tuple<int, int>> GetOffset(IEnumerable<Tuple<int, int>> offsets)
    {
        while (true)
        {
            foreach (var x in offsets)
            {
                yield return x;
            }
        }
    }
}
